<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Methods for real data · MicrobiomeAnalysis.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.044/juliamono.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../readme/">MicrobiomeAnalysis.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../readme/">Home</a></li><li><a class="tocitem" href="../">Manual</a></li><li><span class="tocitem">Tutorial</span><ul><li><a class="tocitem" href="../example1/">SE construction and analysis</a></li><li><a class="tocitem" href="../example2/">MAE construction and manipulation</a></li><li><a class="tocitem" href="../example3/">SE and MAE retrieval</a></li><li><a class="tocitem" href="../example4/">CommunityProfile conversion</a></li><li class="is-active"><a class="tocitem" href>Methods for real data</a><ul class="internal"><li><a class="tocitem" href="#Assays"><span>Assays</span></a></li><li><a class="tocitem" href="#Rowdata-and-Coldata"><span>Rowdata and Coldata</span></a></li><li><a class="tocitem" href="#Filtering"><span>Filtering</span></a></li><li><a class="tocitem" href="#Toolkit-to-manipulate-DataFrames"><span>Toolkit to manipulate DataFrames</span></a></li></ul></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Tutorial</a></li><li class="is-active"><a href>Methods for real data</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Methods for real data</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/JuliaTurkuDataScience/MicrobiomeAnalysis.jl/blob/master/docs/src/example5.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Tutorial-5:-Wrangle-real-data-into-a-SummarizedExperiment"><a class="docs-heading-anchor" href="#Tutorial-5:-Wrangle-real-data-into-a-SummarizedExperiment">Tutorial 5: Wrangle real data into a SummarizedExperiment</a><a id="Tutorial-5:-Wrangle-real-data-into-a-SummarizedExperiment-1"></a><a class="docs-heading-anchor-permalink" href="#Tutorial-5:-Wrangle-real-data-into-a-SummarizedExperiment" title="Permalink"></a></h1><p>Until now, experiments were either simulated, imported from R or converted from a CommunityProfile. Producing instances of SummarizedExperiment from scratch might appear more challenging, as it requires some data analysis skills to wrangle the raw data into the necessary shape. Also, filtering missing values or setting some thresholds could also come into handy. Here, a few possible approaches are explained.</p><h2 id="Assays"><a class="docs-heading-anchor" href="#Assays">Assays</a><a id="Assays-1"></a><a class="docs-heading-anchor-permalink" href="#Assays" title="Permalink"></a></h2><p>It is recommended to store experimental results and bioinformatics output into an Excel workbook or as a group of csv files, which can be easily imported.</p><pre><code class="language-julia hljs">using XLSX, CSV, DataFrames

# read assays stored in same excel and store them into separate variables
first_assay = XLSX.readdata(joinpath(@__DIR__, &quot;assets/RE_assays.xlsx&quot;), &quot;Sheet1&quot;, &quot;B2:E11&quot;) |&gt; Matrix
second_assay = XLSX.readdata(joinpath(@__DIR__, &quot;assets/RE_assays.xlsx&quot;), &quot;Sheet1&quot;, &quot;F2:I11&quot;) |&gt; Matrix

# replace NA values to Julia-native &quot;missing&quot; object class
replace!(first_assay, &quot;NA&quot; =&gt; missing)
replace!(second_assay, &quot;NA&quot; =&gt; missing)

# assemble assays into a single object
assays = OrderedDict{String, AbstractArray}(&quot;first_assay&quot; =&gt; first_assay,
                                            &quot;second_assay&quot; =&gt; second_assay)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">OrderedDict{String, AbstractArray} with 2 entries:
  &quot;first_assay&quot;  =&gt; Any[10482.2 1441.07 6448.97 7737.69; 9401.0 7211.86 4721.28…
  &quot;second_assay&quot; =&gt; Any[10787.3 12275.2 16580.6 2682.84; missing missing missin…</code></pre><h2 id="Rowdata-and-Coldata"><a class="docs-heading-anchor" href="#Rowdata-and-Coldata">Rowdata and Coldata</a><a id="Rowdata-and-Coldata-1"></a><a class="docs-heading-anchor-permalink" href="#Rowdata-and-Coldata" title="Permalink"></a></h2><p>After loading the csv file containing the rowdata, a few operations are needed to make it suitable for feeding into a SummarizedExperiment. For more analytical tips, the <a href="https://dataframes.juliadata.org/stable/">docs on DataFrames.jl</a> and <a href="https://juliaturkudatascience.github.io/MicrobiomeAnalysis.jl/dev/example5/#Toolkit-to-manipulate-DataFrames">this toolkit</a> are recommended.</p><pre><code class="language-julia hljs"># load rowdata from csv file
row_data = CSV.File(joinpath(@__DIR__, &quot;assets/RE_rowdata.csv&quot;)) |&gt; DataFrame

# rename &quot;Column1&quot; to &quot;name&quot;
rename!(row_data, &quot;Column1&quot; =&gt; :name)

# sort by feature number so rowdata order matches assay order
row_data[!, :order] = parse.(Int64, replace.(row_data[!, :name], &quot;bin&quot; =&gt; &quot;&quot;))
sort!(row_data, :order)</code></pre><div class="data-frame"><p>10 rows × 5 columns</p><table class="data-frame"><thead><tr><th></th><th>name</th><th>genus</th><th>species</th><th>native</th><th>order</th></tr><tr><th></th><th title="InlineStrings.String7">String7</th><th title="InlineStrings.String31">String31</th><th title="String">String</th><th title="InlineStrings.String7">String7</th><th title="Int64">Int64</th></tr></thead><tbody><tr><th>1</th><td>bin1</td><td>g__MGIIa-L2</td><td>s__MGIIa-L2 sp013911465</td><td>true</td><td>1</td></tr><tr><th>2</th><td>bin2</td><td>g__UBA3478</td><td>s__UBA3478 sp011525015</td><td>false</td><td>2</td></tr><tr><th>3</th><td>bin3</td><td>g__GCA-2684655</td><td>s__</td><td>true</td><td>3</td></tr><tr><th>4</th><td>bin4</td><td>g__</td><td>s__</td><td>nothing</td><td>4</td></tr><tr><th>5</th><td>bin5</td><td>g__Shewanella</td><td>s__</td><td>true</td><td>5</td></tr><tr><th>6</th><td>bin6</td><td>g__Sphingomonas</td><td>s__Sphingomonas ginsenosidimutans</td><td>true</td><td>6</td></tr><tr><th>7</th><td>bin7</td><td>g__Flavobacterium</td><td>s__Flavobacterium sp002862805</td><td>true</td><td>7</td></tr><tr><th>8</th><td>bin8</td><td>g__Vibrio</td><td>s__Vibrio azureus</td><td>false</td><td>8</td></tr><tr><th>9</th><td>bin9</td><td>g__Polaribacter</td><td>s__</td><td>nothing</td><td>9</td></tr><tr><th>10</th><td>bin10</td><td>g__Alteromonas</td><td>s__Alteromonas sp009811415</td><td>false</td><td>10</td></tr></tbody></table></div><p>The same method can be applied to the coldata. If it is not available or is not relevant for the analysis, it can be substituted with an empty DataFrame containing only the sample names as done below. Then, we gather the components into a SummarizedExperiment.</p><pre><code class="language-julia hljs"># create empty coldata object
col_data = DataFrame(
    name = [&quot;sample$i&quot; for i in 1:4]
);

# assemble into SummarizedExperiment
se = SummarizedExperiment(assays, row_data, col_data)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">10x4 SummarizedExperiment
  assays(2): first_assay second_assay
  rownames: bin1 bin2 ... bin9 bin10
  rowdata(5): name genus ... native order
  colnames: sample1 sample2 sample3 sample4
  coldata(1): name
  metadata(0):</code></pre><h2 id="Filtering"><a class="docs-heading-anchor" href="#Filtering">Filtering</a><a id="Filtering-1"></a><a class="docs-heading-anchor-permalink" href="#Filtering" title="Permalink"></a></h2><p>The most convenient technique to filter a SummarizedExperiment involves the <code>map</code> function combined with either the <code>eachrow</code> or <code>eachcol</code> iterators, depending on whether you want to filter row-wise or column-wise. First, a custom condition, such as <code>x -&gt; mean(x) == 10</code>, should be passed to <code>map</code> together with an iterator across the dimensions of the assay of interest, such as <code>eachcol(assay(se, &quot;my_assay&quot;))</code>. This returns a vector of indices that can be use to filter the rows or columns of the SummarizedExperiment, as shown here:</p><pre><code class="language-julia hljs"># drop rows with missing values in some assay
se2 = dropmissing(se, &quot;second_assay&quot;)

# filter features that have a peak abundance of at least 10000 reads
keep_rows = map(x -&gt; maximum(x) &gt;= 10000, eachrow(assay(se2, &quot;first_assay&quot;)))
se2 = se2[keep_rows, :]

# filter features by that are native to the sampling site
keep_rows = map(x -&gt; x == &quot;true&quot;, rowdata(se2)[!, :native])
se2 = se2[keep_rows, :]

# filter samples that contain mpre than 15000 reads
keep_cols = map(x -&gt; sum(x) &gt; 15000, eachcol(assay(se2, &quot;second_assay&quot;)))
se2 = se2[:, keep_cols]</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">1x1 SummarizedExperiment
  assays(2): first_assay second_assay
  rownames: bin1
  rowdata(5): name genus ... native order
  colnames: sample3
  coldata(1): name
  metadata(0):</code></pre><h2 id="Toolkit-to-manipulate-DataFrames"><a class="docs-heading-anchor" href="#Toolkit-to-manipulate-DataFrames">Toolkit to manipulate DataFrames</a><a id="Toolkit-to-manipulate-DataFrames-1"></a><a class="docs-heading-anchor-permalink" href="#Toolkit-to-manipulate-DataFrames" title="Permalink"></a></h2><p>The following commands come with <em>DataFrames.jl</em> (apart from the last one) and can help you prepare the raw files and ass into a SummarizedExperiment.</p><ul><li>rename!</li><li>replace!</li><li>sort!</li><li>select!</li><li>leftjoin!</li><li>copy</li></ul></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../example4/">« CommunityProfile conversion</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.18 on <span class="colophon-date" title="Wednesday 25 May 2022 03:05">Wednesday 25 May 2022</span>. Using Julia version 1.7.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
