<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Methods for real data · MicrobiomeAnalysis.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.044/juliamono.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../readme/">MicrobiomeAnalysis.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../readme/">Home</a></li><li><a class="tocitem" href="../">Manual</a></li><li><span class="tocitem">Tutorial</span><ul><li><a class="tocitem" href="../example1/">SE construction and analysis</a></li><li><a class="tocitem" href="../example2/">MAE construction and manipulation</a></li><li><a class="tocitem" href="../example3/">SE and MAE retrieval</a></li><li><a class="tocitem" href="../example4/">CommunityProfile conversion</a></li><li class="is-active"><a class="tocitem" href>Methods for real data</a><ul class="internal"><li><a class="tocitem" href="#Assays"><span>Assays</span></a></li><li><a class="tocitem" href="#Rowdata-and-Coldata"><span>Rowdata and Coldata</span></a></li><li><a class="tocitem" href="#filtering"><span>filtering</span></a></li><li><a class="tocitem" href="#Toolkit-to-manipulate-DataFrames"><span>Toolkit to manipulate DataFrames</span></a></li></ul></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Tutorial</a></li><li class="is-active"><a href>Methods for real data</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Methods for real data</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/JuliaTurkuDataScience/MicrobiomeAnalysis.jl/blob/master/docs/src/example5.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Tutorial-5:-Wrangle-real-data-into-a-SummarizedExperiment"><a class="docs-heading-anchor" href="#Tutorial-5:-Wrangle-real-data-into-a-SummarizedExperiment">Tutorial 5: Wrangle real data into a SummarizedExperiment</a><a id="Tutorial-5:-Wrangle-real-data-into-a-SummarizedExperiment-1"></a><a class="docs-heading-anchor-permalink" href="#Tutorial-5:-Wrangle-real-data-into-a-SummarizedExperiment" title="Permalink"></a></h1><p>Until now, experiments were either simulated, imported from R or converted from a CommunityProfile. Producing instances of SummarizedExperiment from scratch might appear more challenging, as it requires some data analysis skills to wrangle the raw data into the necessary shape. Also, filtering missing values or setting some thresholds could also come into handy. Here, a few possible approaches are explained.</p><h2 id="Assays"><a class="docs-heading-anchor" href="#Assays">Assays</a><a id="Assays-1"></a><a class="docs-heading-anchor-permalink" href="#Assays" title="Permalink"></a></h2><p>It is recommended to store experimental results and bioinformatics output into an Excel workbook or as a group of csv files, which can be easily imported.</p><pre><code class="language- hljs">using XLSX, CSV, DataFrames

first_assay = XLSX.readdata(joinpath(@__DIR__, &quot;assets/RE_assays.xlsx&quot;), &quot;Sheet1&quot;, &quot;B2:E11&quot;) |&gt; Matrix
second_assay = XLSX.readdata(joinpath(@__DIR__, &quot;assets/RE_assays.xlsx&quot;), &quot;Sheet1&quot;, &quot;F2:I11&quot;) |&gt; Matrix

replace!(first_assay, &quot;NA&quot; =&gt; missing)
replace!(second_assay, &quot;NA&quot; =&gt; missing)

assays = OrderedDict{String, AbstractArray}(&quot;first_assay&quot; =&gt; first_assay,
                                            &quot;second_assay&quot; =&gt; second_assay)</code></pre><h2 id="Rowdata-and-Coldata"><a class="docs-heading-anchor" href="#Rowdata-and-Coldata">Rowdata and Coldata</a><a id="Rowdata-and-Coldata-1"></a><a class="docs-heading-anchor-permalink" href="#Rowdata-and-Coldata" title="Permalink"></a></h2><pre><code class="language- hljs"># load rowdata from csv file
row_data = CSV.File(joinpath(@__DIR__, &quot;assets/RE_rowdata.csv&quot;)) |&gt; DataFrame

# rename &quot;Column1&quot; to &quot;name&quot;
rename!(row_data, &quot;Column1&quot; =&gt; :name)

# sort by feature number so rowdata order matches assay order
row_data[!, :order] = parse.(Int64, replace.(row_data[!, :name], &quot;bin&quot; =&gt; &quot;&quot;))
sort!(row_data, :order)</code></pre><p>The same method can be applied to the coldata. If it is not available or is not relevant for the analysis, it can be substituted with an empty DataFrame containing only the sample names as done below. Then, we gather the components into a SummarizedExperiment.</p><pre><code class="language- hljs"># create empty coldata object
col_data = DataFrame(
    name = [&quot;sample$i&quot; for i in 1:4]
);

# assemble into SummarizedExperiment
se = SummarizedExperiment(assays, row_data, col_data)</code></pre><h2 id="filtering"><a class="docs-heading-anchor" href="#filtering">filtering</a><a id="filtering-1"></a><a class="docs-heading-anchor-permalink" href="#filtering" title="Permalink"></a></h2><pre><code class="language- hljs"># drop rows with missing values in some assay
se2 = dropmissing(se, &quot;second_assay&quot;)

# filter features that have a peak abundance of at least 10000 reads
keep_rows = map(x -&gt; maximum(x) &gt;= 10000, eachrow(assay(se2, &quot;first_assay&quot;)))
se2 = se2[keep_rows, :]

# filter features by that are native to the sampling site
keep_rows = map(x -&gt; x == &quot;true&quot;, rowdata(se2)[!, :native])
se2 = se2[keep_rows, :]

# filter samples that contain mpre than 15000 reads
keep_cols = map(x -&gt; sum(x) &gt; 15000, eachcol(assay(se2, &quot;second_assay&quot;)))
se2 = se2[:, keep_cols]</code></pre><h2 id="Toolkit-to-manipulate-DataFrames"><a class="docs-heading-anchor" href="#Toolkit-to-manipulate-DataFrames">Toolkit to manipulate DataFrames</a><a id="Toolkit-to-manipulate-DataFrames-1"></a><a class="docs-heading-anchor-permalink" href="#Toolkit-to-manipulate-DataFrames" title="Permalink"></a></h2><p>The following commands come with <em>DataFrames.jl</em> (apart from the last one) and can help you prepare the raw files and ass into a SummarizedExperiment.</p><ul><li>rename!</li><li>replace!</li><li>sort!</li><li>select!</li><li>leftjoin!</li><li>copy</li></ul><p>```</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../example4/">« CommunityProfile conversion</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.18 on <span class="colophon-date" title="Tuesday 24 May 2022 23:39">Tuesday 24 May 2022</span>. Using Julia version 1.7.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
